### PR TITLE
Simulation and test scenarios review

### PR DESCRIPTION
**Pull request by mortelli**
This PR introduces the following changes to the _Simulation and test scenarios_ document:

* Added **expectations** and **restrictions** to some of the test scenarios
* Whitespace and formatting adjustments
* Grammar changes

**All** changes are negotiable; let's discuss if there are disagreements.

Some questions will be included in the form of a review.

### CODE REVIEW

File: doc/incentives/simulation-and-test-scenarios.md [hunk 1]
```diff
@@ -1,87 +1,75 @@
...
+
+Also, some of the tests will need the presence of a simulated blockchain. The network testing track is already setting up the framework which will allow such testing by providing infrastructure accordingly (e.g. Ganache as a blockchain backend).
 
 We can thus roughly distinguish between:
 * Simulations
 * Integration tests
 * Experimentation tests
 
-This document aims at collecting the specification for the set of needed testing.
+This document aims to collect the specification for the set of needed testing.
```
Comment (mortelli) [inline doc/incentives/simulation-and-test-scenarios.md:12]: what are you trying to say with `the set of needed testing`?

File: doc/incentives/simulation-and-test-scenarios.md [hunk 2]
```diff
@@ -1,87 +1,75 @@
...
 ### Cheques correctness
-Trigger multiple thresholds crossing (and thus multiple cheques emission).
-**Description**: Launch x nodes, upload a file, sync, then download from different nodes. 
-**Expectation**: An expected set of nodes has (multiple) cheques. The cumulative amount of cheque value and balances match.
-**Restriction**: Fixed (non random) data and size, use snapshot.
+Trigger multiple threshold crossings (and thus multiple cheque emissions).
 
+**Description**: Launch _x_ nodes, upload a file, sync, then download from different nodes.  
+**Expectation**: An expected set of nodes has (multiple) cheques. The cumulative amount of cheque values and balances match. For every pair of nodes _A_ & _B_ the last received cheque to node _B_ from node _A_ should match the last sent cheque from node _A_ to node _B_.  
+**Restriction**: All nodes should have SWAP enabled and a valid chequebook smart contract deployed. File size should be big enough to trigger cheque emissions, so fixed (non random) data and size should be used. Use snapshot.
```
Comment (mortelli) [inline doc/incentives/simulation-and-test-scenarios.md:29]: how would snapshots be used here?

File: doc/incentives/simulation-and-test-scenarios.md [hunk 3]
```diff
@@ -1,87 +1,75 @@
...
+**Restriction**: All nodes should have SWAP enabled and a valid chequebook smart contract deployed.
 
 ### Failure test
 Test for error scenarios in relation to blockchain and/or oracles.
-**Description**: Deploy x nodes. Test different error scenarios: no blockchain, no oracle, no both, wrong addresses, malfunctioning oracles (if makes sense), etc.
-**Expectation**: All error scenarios are handled in predictable and expected ways. 
-**Restriction**: 
+
+**Description**: Deploy x nodes. Test different error scenarios: no blockchain and/or no oracle, , wrong addresses, malfunctioning oracles (if makes sense), etc.  
+**Expectation**: All error scenarios are handled in predictable and expected ways.
```
Comment (mortelli) [inline doc/incentives/simulation-and-test-scenarios.md:49]: what are the predictable and expected ways? no `panic`?

can you please clarify?

**Review by mortelli — COMMENTED** [review]
left some questions for @holisticode

File: doc/incentives/simulation-and-test-scenarios.md (hunk 1 above)
Comment (holisticode) [reply doc/incentives/simulation-and-test-scenarios.md:12]: That the document specifies the set of tests we need.

File: doc/incentives/simulation-and-test-scenarios.md [hunk 4]
```diff
@@ -1,87 +1,75 @@
...
 
 ### Accounting correctness
-This is a test well suited for a simulation. At no point after a change the correctness of accounting should be broken.
-**Description**: Launch x nodes, upload a file, sync, then download from different nodes, and check balances.
-**Expectation**: All balances should be symmetric with each other (`balanceA == -balanceB && balanceB == -balanceA`)
-**Restriction**: Keep balances below threshold so to not emit cheques 
+This is a test well suited for a simulation. At no point after a change should the correctness of accounting be broken.
+
+**Description**: Launch _x_ nodes, upload a file, sync, then download from different nodes, and check balances.
```
Comment (holisticode) [inline doc/incentives/simulation-and-test-scenarios.md:20]: This is actually not really possible, the current PR #1737 `TestBasicSwapSimulation` covers this but without an actual file upload; including file upload should be done as an integration test. Instead #1737 introduces a simulation with a dummy test protocol, but it basically has the same effect, which is to account for message send/receive.

Will update this myself.

File: doc/incentives/simulation-and-test-scenarios.md [hunk 5]
```diff
@@ -1,87 +1,75 @@
...
 ### Accounting correctness
-This is a test well suited for a simulation. At no point after a change the correctness of accounting should be broken.
-**Description**: Launch x nodes, upload a file, sync, then download from different nodes, and check balances.
-**Expectation**: All balances should be symmetric with each other (`balanceA == -balanceB && balanceB == -balanceA`)
-**Restriction**: Keep balances below threshold so to not emit cheques 
+This is a test well suited for a simulation. At no point after a change should the correctness of accounting be broken.
+
+**Description**: Launch _x_ nodes, upload a file, sync, then download from different nodes, and check balances.  
+**Expectation**: All balances should be symmetric with each other (`balanceA == -balanceB && balanceB == -balanceA`) and at least _some_ of them should be non-zero.  
+**Restriction**: All nodes should have SWAP enabled and a valid chequebook smart contract deployed. All thresholds should be equal and all nodes must be on the same network. Balances should be kept below the payment threshold so no cheques are emitted.
```
Comment (holisticode) [inline doc/incentives/simulation-and-test-scenarios.md:22]: This restriction only applies for upload integration tests as mentioned above, when we have a properly deployed blockchain (ganache) and can deploy chequebooks.

File: doc/incentives/simulation-and-test-scenarios.md [hunk 6]
```diff
@@ -1,87 +1,75 @@
...
+**Restriction**: All nodes should have SWAP enabled and a valid chequebook smart contract deployed. All thresholds should be equal and all nodes must be on the same network. Balances should be kept below the payment threshold so no cheques are emitted.  
 
 ### Cheques correctness
-Trigger multiple thresholds crossing (and thus multiple cheques emission).
-**Description**: Launch x nodes, upload a file, sync, then download from different nodes. 
-**Expectation**: An expected set of nodes has (multiple) cheques. The cumulative amount of cheque value and balances match.
-**Restriction**: Fixed (non random) data and size, use snapshot.
+Trigger multiple threshold crossings (and thus multiple cheque emissions).
 
+**Description**: Launch _x_ nodes, upload a file, sync, then download from different nodes.
```
Comment (holisticode) [inline doc/incentives/simulation-and-test-scenarios.md:27]: This is also not possible as in-process sim. Will adjust. This test is now covered in #1737 as `TestMultiChequeSimulation`

File: doc/incentives/simulation-and-test-scenarios.md (hunk 2 above)
Comment (holisticode) [reply doc/incentives/simulation-and-test-scenarios.md:29]: This does not apply anymore after the experiences with #1737. Will adjust.

File: doc/incentives/simulation-and-test-scenarios.md (hunk 3 above)
Comment (holisticode) [reply doc/incentives/simulation-and-test-scenarios.md:49]: As I said in a meeting, every error scenario requires specific handling, there is no global handling.

Thus, either we add this here (specify all expected error conditions and its mitigation), and/or we do it the way we initially wanted to tackle this, which was through a diagram or error scenario description which @ralph-pichler was going to do.

**Pull request merged**